domain-models-runtime/src/main/java/org/folio/rest/persist/PgUtil.java

```diff
@@ -439,4 +498,137 @@ public static void deleteById(String table, String id,
   public static PostgresClient postgresClient(Context vertxContext, Map<String, String> okapiHeaders) {
     return PostgresClient.getInstance(vertxContext.owner(), TenantTool.tenantId(okapiHeaders));
   }
+  /**
+   * Generate a CQLWrapper object for use by optimization.
+   *
+   * @param query the cql query to parse
+   * @param limit the max objects to return
+   * @param offset the start index of objects to return
+   * @param fields the fields to apply the cql query above to
+   * @return CQLWrapper
+   */
+  public static CQLWrapper createCQLWrapper(
+    String query,
+    int limit,
+    int offset,
+    List<String> fields) throws FieldException {
+
+    CQL2PgJSON cql2pgJson = new CQL2PgJSON(fields);
+
+    return new CQLWrapper(cql2pgJson, query)
+      .setLimit(new Limit(limit))
+      .setOffset(new Offset(offset));
+  }
+  public static class SQLOptimizationException extends Exception {
+    private static final long serialVersionUID = 7772172153233445467L;
+    public SQLOptimizationException(String string) {
+      super(string);
+    }
+  }
+  /**
+   * Generate optimized sql given a specific cql query, tenant, index column name hint and configurable size to hinge the optimization on.
+   *
+   * @param preparedCql the cql query
+   * @param tenantId the tenant used to generate schema location
+   * @param offset start index of objects to return
+   * @param limit max number of objects to return
+   * @param column the index column to use 
+   * @param size the number of rows that determines which method will be used to generate the ultimate result 
+   * @return CQLWrapper
+   * @throws Exception on generation failure
```

Comment:
The original method
https://github.com/folio-org/mod-inventory-storage/blob/v15.1.0/src/main/java/org/folio/rest/impl/InstanceStorageAPI.java#L198
returns false if the preconditions are not satisfied and a regular query witout optimization needs to be executed.
This is a regular program flow.
http://wiki.c2.com/?DontUseExceptionsForFlowControl
http://wiki.c2.com/?PrincipleOfLeastAstonishment
Don't use exceptions for regular program flow.
Instead you can return a special value, for example null or the empty String.